File: tests/cases/masters_beaten_list.py
"""Oracle-diff cases for poketcg/src/engine/masters_beaten_list.asm."""

wMastersBeatenList = 0xD3BB
GUARD_LOW = wMastersBeatenList - 1
GUARD_HIGH = wMastersBeatenList + 10

CONTRACT = {
    "ClearMasterBeatenList": {
        "compare": ("a", "f", "b", "c", "d", "e", "hl"),
        "preserve": ("b", "c", "d", "e", "hl"),
    },
    "AddMasterBeatenToList": {
        "compare": ("a", "f", "b", "c", "d", "e", "hl"),
        "preserve": ("b", "c", "d", "e", "hl"),
    },
}

POISON = {"a": 0xAA, "f": 0xF0, "b": 0xBB, "c": 0xCC, "d": 0xDD, "e": 0xEE, "hl": 0x1234}

CASES = {
    "ClearMasterBeatenList": [
        {"wram": {wMastersBeatenList: b"\xff" * 10}},
        dict(POISON, wram={wMastersBeatenList: b"\xaa\xbb\xcc\xdd\xee\xff\x11\x22\x33\x44"}),
        {"wram": {GUARD_LOW: b"\x5a", wMastersBeatenList: b"\x01" * 10, GUARD_HIGH: b"\xa5"}, "read": {wMastersBeatenList: 10}},
    ],
    "AddMasterBeatenToList": [
        {},
        dict(POISON, wram={wMastersBeatenList: b"\x00" * 10}, read={wMastersBeatenList: 10}),
        {"a": 0x07, "wram": {wMastersBeatenList: b"\x01\x02\x03\x04\x05\x06\x00\x08\x09\x0a"}, "read": {wMastersBeatenList: 10}},
        {"a": 0x07, "wram": {wMastersBeatenList: b"\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a"}, "read": {wMastersBeatenList: 10}},
        {"a": 0x0A, "wram": {wMastersBeatenList: b"\x01\x02\x03\x04\x05\x06\x07\x08\x09\x00\x5a"}, "read": {wMastersBeatenList: 10}},
        {"a": 0x55, "wram": {GUARD_LOW: b"\x5a", wMastersBeatenList: b"\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a", GUARD_HIGH: b"\xa5"}, "read": {GUARD_LOW: 12}},
    ],
}

# >>> factory AddAllMastersToMastersBeatenList
CONTRACT["AddAllMastersToMastersBeatenList"] = {
    "compare": ("a", "f", "b", "c", "d", "e", "hl"),
    "preserve": ("b", "c", "d", "e", "hl"),
}
CASES["AddAllMastersToMastersBeatenList"] = [
    {"wram": {wMastersBeatenList: b"\x00" * 10}, "read": {wMastersBeatenList: 10}},
    dict(POISON, wram={wMastersBeatenList: b"\x00" * 10}, read={wMastersBeatenList: 10}),
    {"wram": {wMastersBeatenList: b"\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a"}, "read": {wMastersBeatenList: 10}},
    {"wram": {GUARD_LOW: b"\x5a", wMastersBeatenList: b"\x03\x00\x00\x00\x00\x00\x00\x00\x00\x00", GUARD_HIGH: b"\xa5"}, "read": {GUARD_LOW: 12}},
]
# <<< factory AddAllMastersToMastersBeatenList

from tests.cases._schema_migration import legacy_to_schema

MUTATIONS = {
    "ClearMasterBeatenList": {
        "source_symbol": "ClearMasterBeatenList",
        "before": "gb_write8(address, 0);",
        "after": "gb_write8(address, 0xFFu);",
        "case_ids": ["ClearMasterBeatenList-0", "ClearMasterBeatenList-1", "ClearMasterBeatenList-2"],
    },
    "AddMasterBeatenToList": {
        "source_symbol": "AddMasterBeatenToList",
        "before": "gb_write8(address, master);",
        "after": "gb_write8(address, (uint8_t)(master ^ 0xFFu));",
        "case_ids": ["AddMasterBeatenToList-1", "AddMasterBeatenToList-2", "AddMasterBeatenToList-4", "AddMasterBeatenToList-5"],
    },
}

SCHEMA2_CASES = legacy_to_schema(CASES, CONTRACT)
# >>> factory-mutation AddAllMastersToMastersBeatenList
MUTATIONS["AddAllMastersToMastersBeatenList"] = {
    "source_symbol": "AddAllMastersToMastersBeatenList",
    "before": "master_id <= MASTERS_BEATEN_COUNT",
    "after": "master_id < MASTERS_BEATEN_COUNT",
    "case_ids": [
        "AddAllMastersToMastersBeatenList-0",
        "AddAllMastersToMastersBeatenList-1",
        "AddAllMastersToMastersBeatenList-2",
        "AddAllMastersToMastersBeatenList-3",
    ],
}
# <<< factory-mutation AddAllMastersToMastersBeatenList

File: src/home/masters_beaten_list.c
#include "home/masters_beaten_list.h"

#include "generated/wram.h"
#include "mem.h"

#define MASTERS_BEATEN_COUNT 10u

uint8_t ClearMasterBeatenList(uint8_t *f)
{
	uint16_t address = wMastersBeatenList_ADDR;
	for (uint8_t count = MASTERS_BEATEN_COUNT; count != 0; --count) {
		gb_write8(address, 0);
		address = (uint16_t)(address + 1u);
	}
	*f = 0xC0u;
	return 0;
}

uint8_t AddMasterBeatenToList(uint8_t a, uint8_t *f)
{
	uint8_t master = a;
	uint16_t address = wMastersBeatenList_ADDR;
	uint8_t count = MASTERS_BEATEN_COUNT;

	while (count != 0) {
		a = gb_read8(address);
		if (a == 0) {
			gb_write8(address, master);
			*f = 0x80u;
			return master;
		}

		uint8_t flags = 0x40u;
		if ((a & 0x0Fu) < (master & 0x0Fu))
			flags |= 0x20u;
		if (a < master)
			flags |= 0x10u;
		if (a == master)
			flags |= 0x80u;
		if (a == master) {
			*f = flags;
			return a;
		}

		address = (uint16_t)(address + 1u);
		--count;
		if (count == 0) {
			*f = (uint8_t)(flags | 0x80u);
			return a;
		}
	}

	*f = 0x80u;
	return a;
}

/* >>> factory AddAllMastersToMastersBeatenList */
/* masters_beaten_list.asm:46-55 */
uint8_t AddAllMastersToMastersBeatenList(uint8_t *f)
{
	uint8_t master_id;
	for (master_id = 1u; master_id <= MASTERS_BEATEN_COUNT; ++master_id) {
		uint8_t discard_f;
		AddMasterBeatenToList(master_id, &discard_f);
	}
	*f = 0xC0u;
	return master_id;
}
/* <<< factory AddAllMastersToMastersBeatenList */
